src/NumericalAlgorithms/LinearSolver/InnerProduct.hpp

```diff
@@ -0,0 +1,55 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+///\file
+/// Defines an inner product for the linear solver
+
+#pragma once
+
+#include <array>
+
+#include "DataStructures/Variables.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/Blaze.hpp"
+#include "Utilities/ForceInline.hpp"
+
+namespace LinearSolver {
+
+/// \ingroup LinearSolverGroup
+/// Implementations of LinearSolver::inner_product.
+namespace InnerProductImpls {
+
+/// The inner product between any types that have a `dot` product
+template <typename Lhs, typename Rhs>
+struct InnerProductImpl {
+  static double apply(const Lhs& lhs, const Rhs& rhs) noexcept {
+    return dot(lhs, rhs);
+  }
+};
+
+/// The inner product between `Variables`
+template <typename LhsTagsList, typename RhsTagsList>
+struct InnerProductImpl<Variables<LhsTagsList>, Variables<RhsTagsList>> {
+  static double apply(const Variables<LhsTagsList>& lhs,
+                      const Variables<RhsTagsList>& rhs) noexcept {
+    const auto size = lhs.size();
+    ASSERT(size == rhs.size(),
+           "The Variables must be of the same size to take an inner product");
+    return ddot_(size, lhs.data(), 1, rhs.data(), 1);
+  }
+};
+
+}  // namespace InnerProductImpls
+
+/*!
+ * \ingroup LinearSolverGroup
+ * \brief An inner product on the vector space w.r.t. which the addition and
+ * scalar multiplication of both `Lhs` and `Rhs` is defined.
+ */
+template <typename Lhs, typename Rhs>
+SPECTRE_ALWAYS_INLINE double inner_product(const Lhs& lhs,
```

Comment:
This doesn't seem linear-solver-specific.  Should it go somewhere more general?

Reply:
The linear solver works under the following assumptions:
- Whatever data is represented by `lhs` and `rhs` can be interpreted as a vector of the same vector space.
- The `*` and `+` operators  of `Lhs` and `Rhs` implement the scalar multiplication and addition in this vector space.
- The `inner_product` is the standard Euclidean dot product in this vector space.

These assumptions are presumably fairly specific to the linear solver, which currently implements the inner product for Blaze vectors and Variables. For example, I don't expect it makes sense to have this `inner_product` for neither `DataVector` nor `Variables` outside of this namespace. If you think otherwise I'd be happy to move it somewhere more general.